EditableTree: sequences

sharptrip opened the pull request:
## Description

This PR introduces EditableTreeSequence and its proxy-based implementation, which replaces unwrapping sequence fields into arrays in EditableTree. This is required to enable editing of EditableTree.

## Reviewer Guidance

I'd like to have some preliminary feedback on the overall concept, and maybe more detailed on the following:
- the way `EditableTreeSequence` and `ProxyTargetSequence` are typed
- the way I am handling empty sequences (by means of `ProxyTarget.isEmpty`)

## Other information or known dependencies

**Step 4** in a series to push code from #12182 to enable editing of EditableTree.

[Step 3](https://github.com/microsoft/FluidFramework/pull/12428)
[Step 2](https://github.com/microsoft/FluidFramework/pull/12306)
[Step 1](https://github.com/microsoft/FluidFramework/pull/12300)

sharptrip committed:
Implement sequence

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTreeSequence.ts:
```diff
@@ -0,0 +1,335 @@
+/**
+ * Sequence field nodes as a sequence of EditableTrees or primitive values (if unwrapped).
+ */
+export interface EditableTreeSequence<T extends EditableTreeOrPrimitive = EditableTreeOrPrimitive>
```
One way I think we could simplify the users of this type, is to remove the generic type parameter. 

Instead of getting a EditableTreeSequence<EditiableTree> or a EditableTreeSequence<UnwrappedEditiableTree>, we can have a single sequence type, which will give you UnwrappedEditiableTrees by default, but we can add an API to get an EditiableTree explicitly (ex: `getWithoutUnwraping(index: number): EditiableTree`).

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTreeSequence.ts:
```diff
@@ -0,0 +1,335 @@
+import { AdaptingProxyHandler, getArrayOwnKeys, keyIsValidIndex } from "./utilities";
+
+/**
+ * Sequence field nodes as a sequence of EditableTrees or primitive values (if unwrapped).
```
If you get an EditableTreeSequence by iterating an EditiableTree node, it might not correspond to a sequence field: you could get one for some other type of field like a value field.

Since all our fields have a schema and field key (including the root, though its field key is a detached sequence id, that still can be used as a LocalFieldKey for it), I think it might make sense to add a FieldSchema and FieldKey fields to this type, and rename it `EditableField` (replacing the existing EditiableField type).

I think that would help simplify the API, while fixing the name of this type at the same time.

NicholasCouri committed:
Moving ElectedClientNotSummarized from sendErrorEvent to sendTelemetryEvent (#12482)

Today, the ElectedClientNotSummarized event is treated as an error and
as the number of hits is dependent to the number of running clients, a
running stress test with 100+ clients will end up generating 100+
errors. This is not necessarily a bug and ends up creating unnecessary
incidents (Ex. [Bug
2241](https://dev.azure.com/fluidframework/internal/_workitems/edit/2241)).
After talking to the OCEs (Navin, Si) and Jatin, we've decided to move
it from an error to a generic telemetry event to reduce the noise from
the system.

ghoshkaj committed:
Change all `export *` in `packages/drivers` to named exports (#12472)

This PR converts all `export *` in `packages/drivers` to named exports.
Related issue: https://github.com/microsoft/FluidFramework/issues/10062
See for the fact that bundle size does not change when the entire repo
is converted:
https://github.com/microsoft/FluidFramework/pull/12321#issue-1400290954.

ghoshkaj committed:
Change all `export *` in `/tools` to named exports (#12463)

This PR converts all `export *` in `/tools` to named exports.
Related issue: https://github.com/microsoft/FluidFramework/issues/10062
See for the fact that bundle size does not change when the entire repo is converted:
https://github.com/microsoft/FluidFramework/pull/12321#issue-1400290954.

ghoshkaj committed:
Change all `export *` in `package/framework` to named exports (#12424)

This PR converts `export *`  in  `package/framework`  to named exports.
Related issue: https://github.com/microsoft/FluidFramework/issues/10062
See for the fact that bundle size does not change when the entire repo is converted:
https://github.com/microsoft/FluidFramework/pull/12321#issue-1400290954.

ghoshkaj committed:
Change `export *` in `experimental/PropertyDDS` to named exports (#12399)

This PR converts default exports in `experimental/PropertyDDS` to named exports.
Related issue: https://github.com/microsoft/FluidFramework/issues/10062
See for the fact that bundle size does not change when the entire repo is converted:
https://github.com/microsoft/FluidFramework/pull/12321#issue-1400290954.

tylerbutler committed:
build: Pass build-tools version correctly (#12485)

The previous change in #11670 was incomplete. It did not properly pass
through the build tools location from the ADO UI all the way to the
right places. This change corrects that.

I tested by manually running a build with the "next" value for the build
tools source, and verified that the prerelease (tagged `@next`) version
was installed.

sashasimic committed:
Azure scenario tests/runner (#12222)

Through @fluidframework/azure-scenario-runner we are exploring
improvements allowing us to build, execute and validate azure-client
scenarios in more flexible and effective manner. At this point in time,
we will mostly use the scenario runner with FRS team to validate service
characteristics. However, as we learn further, we should be able to use
it beyond service validation.

markfields committed:
Update IAgentScheduler to implement IFluidLoadable (#12491)

AgentScheduler exposes handle but not via IFluidLoadable which makes it
hard to access in general.

markfields committed:
Add TestOverride.SweepTimeout (#12419)

## Background

As-is, the Sweep Timeout will always be at least 6 days due to hardcoded
Snapshot Cache limit (5 days) and Buffer (1 day). This means it's
impossible to test the Sweep Timer and the code that runs after
something is SweepReady!

There have been multiple attempts on my part to address this: #11084,
#12044, #12331

Much of the difficulty has arisen from the challenge of aligning driver
policy with GC config. I'm finally punting on that and tracking it as a
separate item
([AB#2238](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/2238)).

## This change

It's pretty simple:

* Persist SweepTimeoutMs in the summary
* For new containers, that can be overridden via the
`"Fluid.GarbageCollection.TestOverride.SweepTimeoutMs"` config setting
* Default behavior for new containers (and for backfilling existing
containers with no value persisted) is SweepTimeout + 6 days (same as it
has been)

jikim-msft committed:
Rerun Publish Steps after Network Error (#12413)

### Description
665: Ability to Rerun Publish Step If There Are Any Network Error
https://dev.azure.com/fluidframework/internal/_workitems/edit/665/

### Cases
1. If Publish Package does not encounter an error continue with the
remaining packages

2. If Publish Package encounters an error
- Network Error: retry for `$maximumRetryIfNetworkError`-amount of time.
If network error is persisted during the entire loop, exit the task in
the pipeline
- Non-Network Error: Such as the current package is already published in
the registry, fail the pipeline and not continue
